src/theory/quantifiers/ceg_t_instantiator.h

```diff
@@ -76,11 +77,65 @@ class EprInstantiator : public Instantiator {
   std::string identify() const { return "Epr"; }
 };
 
+
+// virtual class for model queries
+class BvInverterModelQuery {
+public:
+  BvInverterModelQuery(){}
+  virtual Node getModelValue( Node n ) = 0;
+};
+
+// class for storing information about the solved status
+class BvInverterStatus {
+public:
+  BvInverterStatus() : d_status(0) {}
+  int d_status;
+  std::vector< Node > d_conds;
+};
+
+// inverter class
+// TODO : move to theory/bv/ if generally useful?
+class BvInverter {
+private:
+  std::map< TypeNode, Node > d_solve_var;
+  Node getPathToPv( Node lit, Node pv, Node sv, std::vector< unsigned >& path, std::map< TNode, bool >& visited );
+public:
+  // get dummy fresh variable of type tn, used as argument for sv 
+  Node getSolveVariable( TypeNode tn );
+  // Get path to pv in lit, replace that occurrence by sv and all others by pvs.
+  // e.g. if return value R is non-null, then:
+  //   lit.path = pv
+  //   R.path = sv
+  //   R.path' = pvs for all lit.path' = pv, where path' != path
+  Node getPathToPv( Node lit, Node pv, Node sv, Node pvs, std::vector< unsigned >& path );
+public:
+  // solve for sv in constraint ( (pol ? _ : not) sv_t <k> t ), where sv_t.path = sv
+  Node solve_bv_constraint( Node sv, Node sv_t, Node t, Kind rk, bool pol, std::vector< unsigned >& path,
+                            BvInverterModelQuery * m, BvInverterStatus& status );
+  // solve for sv in lit, where lit.path = sv
+  Node solve_bv_lit( Node sv, Node lit, bool pol, std::vector< unsigned >& path,
+                     BvInverterModelQuery * m, BvInverterStatus& status );
+};
+
 class BvInstantiator : public Instantiator {
+private:
+  BvInverter d_inverter;
+  unsigned d_inst_id_counter;
+  std::map< Node, std::vector< unsigned > > d_var_to_inst_id;
+  std::map< unsigned, Node > d_inst_id_to_term;
+  std::map< unsigned, BvInverterStatus > d_inst_id_to_status;
+private:
+  void processLiteral( CegInstantiator * ci, SolvedForm& sf, Node pv, Node lit, unsigned effort );
 public:
   BvInstantiator( QuantifiersEngine * qe, TypeNode tn ) : Instantiator( qe, tn ){}
   virtual ~BvInstantiator(){}
+  void reset( CegInstantiator * ci, SolvedForm& sf, Node pv, unsigned effort );
+  bool hasProcessEquality( CegInstantiator * ci, SolvedForm& sf, Node pv, unsigned effort ) { return true; }
```

Comment:
I find it helpful when methods overwriting super class methods are declared `virtual` in the subclasses as well.

Reply:
This is a matter of style, I dislike having virtual classes that are more than one deep, but I could see some arguments in favor of specifying virtual anyways.

Reply:
This is definitely a matter of style. However, I am not quite sure what the relationship with not having virtual classes that are more than one deep is. Adding the `virtual` keyword to methods overwriting methods in the superclass is a noop and would be for documentation purposes only.